pkg/controller/instance/instance_controller.go

```diff
@@ -543,19 +516,112 @@ func isUpgradePlan(planName string) bool {
 	return planName == v1beta1.DeployPlanName || planName == v1beta1.UpgradePlanName
 }
 
-// getPlanToBeExecuted returns name of the plan that should be executed
-func getPlanToBeExecuted(i *v1beta1.Instance, ov *v1beta1.OperatorVersion) (*string, error) {
+func areWebhooksEnabled() bool {
+	return strings.ToLower(os.Getenv("ENABLE_WEBHOOKS")) == "true"
+}
+
+// getPlanInProgress method returns current plan that is in progress. As long as we don't enforce webhooks
+// (ENABLE_WEBHOOKS=true) we have TWO WAYS of deciding that:
+//
+// 1. WITHOUT the webhook, we have the old logic which iterates over all existing PlanStatuses, searching for
+//    the first one with Status == InProgress. It might happen that multiple plans are in progress (e.g. cleanup and deploy)
+// 2. WITH the webhook, current plan lives in the Spec.PlanExecution.PlanName field
+func getPlanInProgress(i *v1beta1.Instance) *kudov1beta1.PlanStatus {
+	if areWebhooksEnabled() {
+		return i.PlanStatus(i.Spec.PlanExecution.PlanName)
+	}
+	return i.GetPlanInProgress()
+}
+
+// newExecutionPlan method returns a new execution plan (if exists) or nil otherwise. As long as we don't enforce webhooks
+// (ENABLE_WEBHOOKS=true) we have TWO WAYS of deciding which plan has to executed next:
+//
+// 1. WITHOUT the webhook, we have the old logic which tries to infer the change (parameter update, new OperatorVersion etc.)
+//    by diffing current state with the one from the snapshot.
+// 2. WITH the webhook, instance admission webhook has already decided on the plan and the result is in the
+//    Spec.PlanExecution.PlanName field.
+func newExecutionPlan(i *v1beta1.Instance, ov *v1beta1.OperatorVersion) (plan *string, err error) {
+	if areWebhooksEnabled() {
+		if plan, err = fetchNewExecutionPlan(i, ov); plan != nil && err == nil {
+			log.Printf("InstanceController: Fetched new execution plan '%s' from the spec for instance %s/%s", kudo.StringValue(plan), i.Namespace, i.Name)
+		}
```

Comment:
Not sure I like this combination of `if` and named return values here.

I'd prefer to have either not have named return values or at least have the if split out from the setting of the variables:
```
plan, err = fetchNewExecutionPlan(i, ov)
if plan != nil && err == nil {
  log.Printf(...)
}
```

Took me longer than expected to parse what's going on here, maybe i'm just not used to named return values

Reply:
I admit I was trying to save some SLOCs here. Without named return values you have at least 2 (!) SLOCs more 😃